projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/op/arith/floating/LLVMArithmeticFactory.java

```diff
@@ -0,0 +1,190 @@
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this list of
+ * conditions and the following disclaimer in the documentation and/or other materials provided
+ * with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to
+ * endorse or promote products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.nodes.op.arith.floating;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropException;
+import com.oracle.truffle.llvm.nodes.op.arith.floating.LLVMArithmeticFactoryFactory.CachedAddNodeGen;
+import com.oracle.truffle.llvm.nodes.op.arith.floating.LLVMArithmeticFactoryFactory.CachedDivNodeGen;
+import com.oracle.truffle.llvm.nodes.op.arith.floating.LLVMArithmeticFactoryFactory.CachedMulNodeGen;
+import com.oracle.truffle.llvm.nodes.op.arith.floating.LLVMArithmeticFactoryFactory.CachedRemNodeGen;
+import com.oracle.truffle.llvm.nodes.op.arith.floating.LLVMArithmeticFactoryFactory.CachedSubNodeGen;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMArithmetic;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMArithmetic.LLVMArithmeticOpNode;
+
+public abstract class LLVMArithmeticFactory {
+    public static LLVMArithmeticOpNode createAddNode() {
+        return CachedAddNodeGen.create();
+    }
+
+    public static LLVMArithmeticOpNode createSubNode() {
+        return CachedSubNodeGen.create();
+    }
+
+    public static LLVMArithmeticOpNode createMulNode() {
+        return CachedMulNodeGen.create();
+    }
+
+    public static LLVMArithmeticOpNode createDivNode() {
+        return CachedDivNodeGen.create();
+    }
+
+    public static LLVMArithmeticOpNode createRemNode() {
+        return CachedRemNodeGen.create();
+    }
+
+    abstract static class CachedAddNode extends LLVMArithmeticOpNode {
+        @Override
+        public boolean canCompute(LLVMArithmetic x, LLVMArithmetic y) {
+            return true;
```

Comment:
shouldn't that be `x instanceof LLVMArithmetic && x.getClass().equals(y.getClass())`?